v3/newrelic/tracing_test.go

```diff
@@ -842,3 +842,82 @@ func TestBetterCAT_SetTraceAndTxnIDs(t *testing.T) {
 		}
 	}
 }
+
+func Test_truncateSpanAttribute(t *testing.T) {
+	tests := []struct {
+		name string // description of this test case
+		// Named input parameters for target function.
+		value                 string
+		maxLengthWithEllipsis int
+		want                  string
+	}{
+		{
+			name:                  "Length of value is less than maxLengthWithEllipsis",
+			value:                 "SELECT * FROM table",
+			maxLengthWithEllipsis: 25,
+			want:                  "SELECT * FROM table",
+		},
+		{
+			name:                  "Length of value is equal to maxLengthWithEllipsis",
+			value:                 "SELECT * FROM table WHERE",
+			maxLengthWithEllipsis: 25,
+			want:                  "SELECT * FROM table WHERE",
+		},
+		{
+			name:                  "Length of value is less than maxLengthWithEllipsis with character larger than a byte",
+			value:                 "SELECT * FROM tablé",
+			maxLengthWithEllipsis: 25,
+			want:                  "SELECT * FROM tablé",
+		},
+		{
+			name:                  "Length of value is equal than maxLengthWithEllipsis with character larger than a byte",
+			value:                 "SELECT * FROM tablé WHERE",
+			maxLengthWithEllipsis: 25,
+			want:                  "SELECT * FROM tablé W...",
+		},
+		{
+			name:                  "Length of value is longer than maxLengthWithEllipsis",
+			value:                 "SELECT * FROM table WHERE condition=truncated",
+			maxLengthWithEllipsis: 25,
+			want:                  "SELECT * FROM table WH...",
+		},
+		{
+			name:                  "Length of value is longer than maxLengthWithEllipsis with character larger than a byte",
+			value:                 "SELECT * FROM tablé WHERE condition=truncated",
+			maxLengthWithEllipsis: 25,
+			want:                  "SELECT * FROM tablé W...",
+		},
+		{
+			name:                  "Length of value is longer than maxLengthWithEllipsis with character larger than a byte in section that gets truncated",
+			value:                 "SELECT * FROM tablé WHERE condition=truncatéd",
+			maxLengthWithEllipsis: 25,
+			want:                  "SELECT * FROM tablé W...",
+		},
+		{
+			name:                  "Length of value is less than maxLengthWithEllipsis and all characters larger than a byte",
+			value:                 "éééé",
+			maxLengthWithEllipsis: 5,
+			want:                  "é...",
+		},
+		{
+			name:                  "Length of value is equal to maxLengthWithEllipsis and all characters larger than a byte",
+			value:                 "ééééé",
+			maxLengthWithEllipsis: 5,
+			want:                  "é...",
+		},
+		{
+			name:                  "Length of value is longer than maxLengthWithEllipsis and all characters larger than a byte",
+			value:                 "éééééé",
+			maxLengthWithEllipsis: 5,
```

Comment:
Could you add one test where the `maxLengthWithEllipsis` value is something like `6` (or any other nearby even number)? All the tests above are exercising odd numbers with multibyte string characters, I think it would be good to verify behavior for even number limits as well.

Reply:
Good catch.  This actually uncovered a bug that will truncate a character to only the max size - 3 bytes which could leave an incomplete character

Reply:
Follow up! Pushed up the changes and re-used the existing string concatenation that we have `stringLengthByteLimit` to properly concatenate multi-byte strings.  Odd thing though is when we call `addString` for the `db.statement` specifically, we actually call that function twice now (since it is called within `truncateSpanAttribute`.  This shouldn't be too much of a problem since it won't truncate twice because the length would be less than the max but I just think it's a little awkward.  Any ideas on a different approach?

Reply:
Documenting what we discussed here for posterity- A possible solution is to use the `key` string value passed to `addString` to gate DB vs non-DB behavior, which should eliminate the need for multiple truncate calls since the truncate logic can be handled directly in the `addString` function itself (for DB specifically)